BootloaderCommonPkg/Library/UsbInitLib/UsbInitLib.c

```diff
@@ -109,7 +110,16 @@ InitUsbDevices (
         (Class[2] == PCI_CLASS_SERIAL))) {
       // Enable XHCI controller
       MmioOr8 (PcieAddress + PCI_COMMAND_OFFSET, EFI_PCI_COMMAND_MEMORY_SPACE | EFI_PCI_COMMAND_BUS_MASTER);
-      XhciMmioBase = (UINTN) MmioRead64 (PcieAddress + PCI_BASE_ADDRESSREG_OFFSET) & ~0xF;
+
+      //
+      // Read high 32-bit BAR only if 64-bit operation and 64-bit BAR memory space
+      //
+      XhciMmioBase = (UINT32)MmioRead32 (PcieAddress + PCI_BASE_ADDRESSREG_OFFSET);
+      BarType = (UINT8)XhciMmioBase & 0xF;
+      XhciMmioBase &= (UINT32)(~BarType);
+      if (IS_X64 && ((BarType & 0x04) != 0)) {

```

Comment:
Do we really need IX_X64 check here ?
in 32 bit mode I think it is also valid to read the full 64bit bar.  The higher 32 bit should be always 0 though.

Reply:
This was to avoid unnecessary high 32-bit BAR reading in IA32. Let me declare XhciMmioBase as UINT64 instead of UINTN and remove IS_X64.